### PR TITLE
add public_accessors option

### PR DESCRIPTION
**Pull request by morgvanny**
following up on #4 

hopefully this makes sense! I didn't have a problem with the existing implementation, but when I tried to add the feature, I ended up making a huge mess. cleaning that up, I found some ways to simplify the logic in general I think

### CODE REVIEW

File: lib/smart_init/main.rb [hunk 1]
```diff
@@ -11,60 +11,46 @@ def is_callable(opts={})
...
+      public_readers = required_attrs
+      public_accessors = required_attrs if public_accessors == true
+    else
+      public_readers += public_accessors
+    end
 
     define_method :initialize do |*parameters|
-      parameters = [{}] if parameters == []
-      unless parameters.first.is_a?(Hash)
+      parameters == [] ? parameters = {} : parameters = parameters.first
```
Comment (pawurb) [inline lib/smart_init/main.rb]: maybe:

```
       parameters = parameters == [] ? {} : parameters.first
```

File: lib/smart_init/main.rb [hunk 2]
```diff
@@ -11,60 +11,46 @@ def is_callable(opts={})
...
-        el.map { |k, v| { k => v } }
-      else
-        el
-      end
-    end.flatten
-
-    public_readers = attributes.select(&public_readers_filter)
-    attributes.delete_if(&public_readers_filter)
-    required_attrs = attributes.select { |el| el.is_a?(Symbol) }
+  def initialize_with_hash(*required_attrs, **hash)
```
Comment (pawurb) [inline lib/smart_init/main.rb]: hash is a global ruby method please rename to prevent shadowing

**Review by pawurb — CHANGES_REQUESTED** [review]
Looking good, just small changes needed. And thanks for a refactor. Please also add info to readme.

File: lib/smart_init/main.rb (hunk 1 above)
Comment (pawurb) [reply lib/smart_init/main.rb]: or  even more readable:

```
parameters = if parameters == []
  {}
else
  parameters.first
end
```

**Comment by morgvanny** [comment]
Thanks for the prompt review! I think I've resolved each of those issues now

**Comment by pawurb** [comment]
Thanks!

**Pull request merged**